packages/query-core/src/utils.ts

```diff
@@ -245,50 +245,104 @@ export function partialMatchKey(a: any, b: any): boolean {
   return false
 }
 
+const hasOwn = Object.prototype.hasOwnProperty
+
 /**
  * This function returns `a` if `b` is deeply equal.
  * If not, it will replace any deeply equal children of `b` with those of `a`.
  * This can be used for structural sharing between JSON values for example.
  */
 export function replaceEqualDeep<T>(a: unknown, b: T): T
-export function replaceEqualDeep(a: any, b: any): any {
+export function replaceEqualDeep(a: unknown, b: unknown): any {
   if (a === b) {
     return a
   }
 
-  const array = isPlainArray(a) && isPlainArray(b)
-
-  if (array || (isPlainObject(a) && isPlainObject(b))) {
-    const aItems = array ? a : Object.keys(a)
-    const aSize = aItems.length
-    const bItems = array ? b : Object.keys(b)
-    const bSize = bItems.length
-    const copy: any = array ? [] : {}
-    const aItemsSet = new Set(aItems)
+  const aIsArr = isPlainArray(a)
+  const bIsArr = isPlainArray(b)
 
+  // both are arrays
+  if (aIsArr && bIsArr) {
+    const aSize = a.length
+    const bSize = b.length
+    const copy: Array<unknown> = new Array(bSize)
     let equalItems = 0
 
     for (let i = 0; i < bSize; i++) {
-      const key = array ? i : bItems[i]
+      const aItem = a[i]
+      const bItem = b[i]
+
+      // most common case (strict equality)
+      if (aItem === bItem) {
+        copy[i] = aItem
+        if (i < aSize) equalItems++
+        continue
+      }
+
+      // either item is not an array or object
       if (
-        ((!array && aItemsSet.has(key)) || array) &&
-        a[key] === undefined &&
-        b[key] === undefined
+        aItem === null ||
+        bItem === null ||
+        typeof aItem !== 'object' ||
+        typeof bItem !== 'object'
       ) {
-        copy[key] = undefined
-        equalItems++
-      } else {
-        copy[key] = replaceEqualDeep(a[key], b[key])
-        if (copy[key] === a[key] && a[key] !== undefined) {
-          equalItems++
-        }
+        copy[i] = bItem
+        continue
       }
+
+      const v = replaceEqualDeep(aItem, bItem)
+      copy[i] = v
+      if (v === aItem) equalItems++
     }
 
     return aSize === bSize && equalItems === aSize ? a : copy
   }
 
-  return b
+  // only 1 is an array
+  if (aIsArr || bIsArr) {
+    return b
+  }
+
+  // at least 1 is not an object
+  if (!isPlainObject(a) || !isPlainObject(b)) {
+    return b
+  }
```

Comment:
could the “only 1 is an array” case not be removed because we would also go into the “at least 1 is not an object” case and also `return b` there?

all tests still pass if I just remove that check, and we could then also inline `aIsArr` and `bIsArr` because they are only used once then ...